product_state/models/product_template.py

```diff
@@ -1,18 +1,75 @@
 # Copyright 2017 ACSONE SA/NV (<http://acsone.eu>)
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
-from odoo import models, fields
+from odoo import api, models, fields
+
+
+class ProductState(models.Model):
+    _name = 'product.state'
+    _description = "Product State"
+    _order = 'sequence'
+
+    name = fields.Char('State Name', required=True, translate=True)
+    code = fields.Char('State Code', required=True)
+    sequence = fields.Integer('Sequence', help="Used to order the States")
+    description = fields.Text(translate=True)
+    product_ids = fields.One2many(
+        'product.template',
+        'product_state_id',
+        string='State Products',
+    )
+    products_count = fields.Integer(
+        string='Number of products',
+        compute='_compute_products_count',
+    )
+
+    _sql_constraints = [
+        ('product_state_code_unique', 'UNIQUE(code)', 'Product State Code must be unique.')
+        ]
+
+    @api.multi
+    @api.depends('product_ids')
+    def _compute_products_count(self):
+        data = self.env["product.template"].read_group([
+            ('product_state_id', 'in', self.ids),
+        ], ['product_state_id'], ['product_state_id'])
+        mapped_data = {
+            record['product_state_id'][0]: record['product_state_id_count']
+            for record in data
+        }
+        for state in self:
+            state.products_count = mapped_data.get(state.id, 0)
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
-    state = fields.Selection(selection=[
-        ('draft', 'In Development'),
-        ('sellable', 'Normal'),
-        ('end', 'End of Lifecycle'),
-        ('obsolete', 'Obsolete')],
-        string='Status',
-        default='sellable',
-        index=True
+    product_state_id = fields.Many2one(
+        'product.state',
+        string='State',
+        default='product_state_draft',
+        help='Select a state for this product'
+    )
+    state = fields.Char(
+        string='Product Status',
+        index=True,
+    )
+    product_state_id = fields.Many2one(
+        'product.state',
+        string='State',
+        help='Select a state for this product',
+        compute="_compute_product_state",
+        inverse="_inverse_product_state",
+        store=True,
     )
+
+    @api.depends('state')
+    def _compute_product_state(self):
+        ProductState = self.env['product.state']
+        for product_tmpl in self:
+            state = ProductState.search([('name', '=', product_tmpl.state)], limit=1)
+            product_tmpl.product_state_id = state.id
```

Comment:
1. You forgot change name -> code in domain
2. You asked me about default value for product_state_id, so it can be done here:
<pre>
@api.depends('state')
def _compute_product_state(self):
    ProductState = self.env['product.state']
    for product_tmpl in self:
        product_state = ProductState
        if product_tmpl.state:
            product_state = ProductState.search([('name', '=', product_tmpl.state)], limit=1)
        if not product_state:
            product_state = ProductState.search([], limit=1)
        product_tmpl.product_state_id = product_state.id
</pre>